core/src/protocols_handler/mod.rs

```diff
@@ -246,6 +236,69 @@ pub trait ProtocolsHandler {
     }
 }
 
+pub struct ListenProtocol<TUpgrade, TSubstream> {
+    upgrade: TUpgrade,
+    timeout: Duration,
+    _marker: std::marker::PhantomData<TSubstream>
+}
+
+impl<TUpgrade, TSubstream> Clone for ListenProtocol<TUpgrade, TSubstream>
+where
+    TUpgrade: Clone
+{
+    fn clone(&self) -> Self {
+        ListenProtocol {
+            upgrade: self.upgrade.clone(),
+            timeout: self.timeout,
+            _marker: std::marker::PhantomData
+        }
+    }
+}
+
+impl<TUpgrade, TSubstream> ListenProtocol<TUpgrade, TSubstream> {
+    pub fn new(upgrade: TUpgrade) -> ListenProtocol<TUpgrade, TSubstream> {
```

Comment:
Nit: I think we should have an `impl From<TUpgrade> for ListenProtocol` that does the same thing as `new`.